### ISSUE TITLE
Implementing a simple PassThrough filter for PointCloud2

### ISSUE THREAD

**Issue by torydebra** [ctu-vras/sensor_filters#3]
Hi,
I am trying to implement a simple passthrough filter for pointcloud2 using the sensor_filters package  
(please tell me if it is already available, at the moment I only found the [pcl_ros one](http://wiki.ros.org/pcl_ros/Tutorials/PassThrough%20filtering) which as you mention is not compatible with the `ros filters`)

I have some doubts:
- I saw your [PointCloud2FilterChainBase](https://github.com/ctu-vras/sensor_filters/blob/master/include/sensor_filters/PointCloud2FilterChainBase.h) class, should I derive this or start directly from the [example](https://github.com/ctu-vras/sensor_filters/blob/master/include/sensor_filters/PointCloud2FilterChainBase.h)? 

- Should I "templatize" the FilterChainNodelet<> with a `sensor_msgs::PointCloud2` ? Why not using directly `pcl::PointCloud<PointXYZ>`? ROS can subscribe directly to this type (eg  `void cloudClbk(const pcl::PointCloud<PointXYZ>::ConstPtr& msg);`), and by doing so we should be able to directly use the pcl methods like:
  ```cpp
    bool PassThroughFilter::filter(const pcl::PointCloud<PointXYZ>& msgIn, pcl::PointCloud<PointXYZ>& msgOut) {
       pcl::PassThrough<pcl::PointXYZ> pass;
       pass.setInputCloud (msgIn);
       pass.setFilterFieldName ("z");
       pass.setFilterLimits (0.0, 1.0);
      //pass.setNegative (true);
       pass.filter (msgOut);
    }
  ```
  without having to convert from `sensor_msgs::PointCloud2` to `pcl::PointCloud<PointXYZ>` which sounds annoying since it seems we have to pass through a third type `pcl::PCLPointCloud2` [example here](https://answers.ros.org/question/136916/conversion-from-sensor_msgspointcloud2-to-pclpointcloudt/)

Thanks for any help!

**Comment by peci1** [comment]
Hi, the approach would be a bit different.

What this package provides are the "runners" (i.e. filter chain implementations) that are able to load and "execute" filters. It does not provide any filters itself (except for the generic ChangeHeaderFilter).

In order to implement a pointcloud filter, you would need to inherit from `filters::FilterBase<sensor_msgs::PointCloud2>` (i.e. package https://wiki.ros.org/filters).

For some time I also wanted to create a "mirror" of the pcl_ros package, but providing this kind of filters instead of nodelets. But I haven't found time to do it yet.

If you do, let me know, I'll be happy to add a link to this package's readme.

---

Regarding the `pcl::PointCloud<PointXYZ>` type, you are right it might be easier to also provide a filter chain working on this types. However, it would be a bit side-stepping. What happens on the background, if you subscribe to this type, is that ROS actually subscribes to `sensor_msgs::PointCloud2` and converts it for you to the PCL type. Where some effort could be saved would be when you have multiple filters and each of them wants to work on the PCL type. It could be done, but I don't think this package is the right place - it would add a dependency on pcl_ros, which not everybody wants... I'd like to keep this a generic package for all sensor message types. Maybe if you're creating the pcl_ros filters, you can also add the filterchain nodes and nodelets for the PCL types there... It should be easy by re-using this package's templates. Just try to use it as if `pcl::PointCloud<pcl::PointXYZ>` were just another sensor_msgs type - I hope that could work.

**Comment by torydebra** [comment]
Ok I misunderstood how these packages works, thanks for the clarification. 
I have implemented a passthrough (https://github.com/torydebra/point_cloud2_filters) based on the `filters::FilterBase<sensor_msgs::PointCloud2>` class that seems to work. 

I used the `sensor_msgs::PointCloud2` and not the `pcl::PointCloud<PointXYZ>` type, since at the moment I want to use the passthrough with the `robot_body_filter` which uses the `sensor_msgs::PointCloud2` too.
